README.md
# built_in_keyboard

A flutter package to add built-in keyboard instead of system keyboards. This implementation relies on a textInputField and interacts with it through a `TextEditingController`. 

<p align="center">
  <img src="https://raw.githubusercontent.com/Open-Segmentation-Systems/BuiltInKeyboard/main/BIKST.png" width="55%">
</p>

## Usage
First of all, add the package to your dependencies and run `flutter pub get`.
```yaml
dependencies:
  built_in_keyboard:
    git:
      url: https://github.com/Open-pi/builtinkeyboard.git
      ref: main
```

With the `BuiltInKeyboard` widget, you can easily change the layout of the keyboard, the look and feel of the keys, etc.

Here are some attributes that you could change. keep in mind that the controller is required to make the keyboard work correctly.
```dart
BuiltInKeyboard(
  controller: this.textController, // required
  language: Language.DE, // optional
  layout: Layout.QWERTZ, // optional, BUT MUST be checked if you modify the language!
  enableSpaceBar: true, // optional, add a spacebar
  enableBackSpace: true, // optional, add a backspace button
  enableCapsLock: true, // optional, add a Caps Lock button
  borderRadius: BorderRadius.circular(8.0), // changes the border radius of the keys
  letterStyle: TextStyle(fontSize: 25, color: Colors.black) // styles the text inside the keys
)
```
### Additional options and flags
Option | Required | By default | Description
--- | --- | --- | ---
**controller** | yes | - | `TextEditingController`
**language** | no | Language.EN | the language of the keyboard
**layout** | no | Layout.QWERTY | the layout of the keyboard
**height** | no | - | height of keys
**width** | no | - | width of keys
**spacing** | no | 8.0 | the spacing between each row
**borderRadius** | no | - | the radius of the keys
**letterStyle** | no | fontSize: 25, color: black | `TextStyle` of the letters
**color** | no | deepOrange | color of the keys (keyboard color)
**enableSpacebar** | no | false | adds a spacebar
**enableBackSpace** | no | true | adds a backspace button
**enableCapsLock** | no | false | adds a Caps Lock button
**enableAllUppercase** | no | false | makes the keyboard uppercase
**enableLongPressUppercase** | no | false | writes an uppercase when long pressing on the keys
**highlightColor** | no | - | color when pressed
**splashColor** | no | - | color when pressed (material style)

## Contribution
### Languages
To add a new language or a keyboad layout to the package you only need to modify the language.dart file. The following steps will show you how to do that.

1. Add the short name form of your new language to the `Language` enum.
```dart
enum Language {
  EN,
  FR,
  DE,
  <your language>,
}
```
2. Add your layout name to the `Layout` enum if not present.
```dart
enum Layout {
  QWERTY,
  QWERTZ,
  AZERTY,
  <your layout>,
}
```
3. Create a new map variable called `<language name>Config`. The keys of this map will be the layouts from the `Layout` enum (e.g: Layout.QWERTY), and the values will be maps with types `<String, String>` that contain the core configuration. These latter maps must contain four fields o:
   1. `layout`: The full text layout of the keyboard.
   2. `horizontalSpacing`: A floating number which will represent the spaceing between each key. 
   3. `topLength`: The lenght of the top/first row of keys. In other words, the number of keys to display in the top row.
   4. `middleLength`: The number of keys in the middle row.

    (The bottom/last row will just take the remaining keys).

    example:
    ```dart
    var frenchConfig = {
      Layout.QWERTY: <String, String>{
        'layout': 'qwertyuiopasdfghjklzxcvbnm',
        'horizontalSpacing': '6.0', // space between keys
        'topLength': '10', // number of keys
        'middleLength': '9', // number of keys
      },
      Layout.AZERTY: <String, String>{
        'layout': 'azertyuiopqsdfghjklmwxcvbn',
        'horizontalSpacing': '6.0', // space between keys
        'topLength': '10', // number of keys
        'middleLength': '9', // number of keys
      },
    };
    ```
4. And finally your new language config map to the `languageConfig` map, with the key as the short language name from the `Language` enum.
```dart
var languageConfig = {
  Language.EN: englishConfig,
  Language.FR: frenchConfig,
  Language.DE: germanConfig,
  Language.<short language name>: <language name>Config,
};
```

Feel free to also modify or add new things to existing languges.




CHANGELOG.md
## [0.2.0]

* Add null safety support.
* Add multi-languages support.

## [0.1.0]

* First 0.1 release.
* Added spacebar button.
* Added caps lock button.
* Made keyboard size adaptive to the screen size.
* Added more parameters.
* Added an example.

## [0.0.1 - 0.0.2]

* Initial release.




lib/language.dart
enum Language {
  EN,
  FR,
  DE,
  AR,
}

enum Layout {
  QWERTY,
  QWERTZ,
  AZERTY,
  ARABIC,
}

var languageConfig = {
  Language.EN: englishConfig,
  Language.FR: frenchConfig,
  Language.DE: germanConfig,
  Language.AR: arabicConfig,
};

// Languages Configurations

var englishConfig = {
  Layout.QWERTY: <String, String>{
    'layout': 'qwertyuiopasdfghjklzxcvbnm',
    'horizontalSpacing': '6.0',
    'topLength': '10',
    'middleLength': '9',
  },
  Layout.QWERTZ: <String, String>{
    'layout': 'qwertzuiopasdfghjklyxcvbnm',
    'horizontalSpacing': '6.0',
    'topLength': '10',
    'middleLength': '9',
  },
};

var frenchConfig = {
  Layout.QWERTY: <String, String>{
    'layout': 'qwertyuiopasdfghjklzxcvbnm',
    'horizontalSpacing': '6.0',
    'topLength': '10',
    'middleLength': '9',
  },
  Layout.AZERTY: <String, String>{
    'layout': 'azertyuiopqsdfghjklmwxcvbn',
    'horizontalSpacing': '6.0',
    'topLength': '10',
    'middleLength': '9',
  },
};

var germanConfig = {
  Layout.QWERTY: <String, String>{
    'layout': 'qwertyuiopüasdfghjklöäzxcvbnmß',
    'horizontalSpacing': '2.5',
    'topLength': '11',
    'middleLength': '11',
  },
  Layout.QWERTZ: <String, String>{
    'layout': 'qwertzuiopüasdfghjklöäyxcvbnmß',
    'horizontalSpacing': '2.5',
    'topLength': '11',
    'middleLength': '11',
  },
};

var arabicConfig = {
  Layout.ARABIC: <String, String>{
    'layout': 'ثةورزدذطظكمنتالبيسشجحخهعغفقصض',
    'horizontalSpacing': '2.8',
    'topLength': '11',
    'middleLength': '10',
  },
};


example/lib/main.dart
import 'package:built_in_keyboard/language.dart';
import 'package:flutter/material.dart';
import 'package:built_in_keyboard/built_in_keyboard.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Built In Keyboard Test',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Built In Keyboard Test Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  @override
  Widget build(BuildContext context) {
    TextEditingController textController = new TextEditingController();
    return Scaffold(
      appBar: AppBar(
        title: Text('Built In Keyboard'),
        centerTitle: true,
        backgroundColor: Colors.deepOrange,
      ),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Spacer(),
          Padding(
            padding: const EdgeInsets.fromLTRB(10.0, 10.0, 10.0, 0.0),
            child: TextFormField(
              controller: textController,
              readOnly: true,
              keyboardType: TextInputType.name,
              style: TextStyle(
                fontWeight: FontWeight.bold,
                color: Colors.indigo[900],
                fontSize: 22.0,
              ),
              decoration: InputDecoration(
                contentPadding: EdgeInsets.fromLTRB(21.0, 0.0, 0.0, 0.0),
                filled: true,
                enabledBorder: OutlineInputBorder(
                  borderRadius: BorderRadius.circular(25.0),
                  borderSide: BorderSide(width: 3.0, color: Colors.deepOrange),
                ),
                focusedBorder: OutlineInputBorder(
                  borderRadius: BorderRadius.circular(25.0),
                  borderSide: BorderSide(width: 3.0, color: Colors.deepOrange),
                ),
                hintText: 'try it...',
                hintStyle: TextStyle(fontSize: 20, fontWeight: FontWeight.w700),
                floatingLabelBehavior: FloatingLabelBehavior.always,
              ),
            ),
          ),
          Spacer(),
          Padding(
            padding: EdgeInsets.only(bottom: 3),
            child: BuiltInKeyboard(
              language: Language.EN,
              layout: Layout.QWERTY,
              borderRadius: BorderRadius.circular(8),
              controller: textController,
              enableLongPressUppercase: true,
              enableSpaceBar: true,
              enableBackSpace: true,
              enableCapsLock: true,
            ),
          ),
        ],
      ),
    );
  }
}


lib/builtInKeyboard.dart
library built_in_keyboard;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

class BuiltInKeyboard extends StatefulWidget {
  // layoutType of the keyboard
  final String layoutType;

  // TextStyle of the letters in the keys (fontsize, fontface)
  final TextStyle letterStyle;

  // Border radius of the keys
  final BorderRadius borderRadius;

  // Color of the keys
  final Color color;

  // The color displayed when the key is pressed
  final Color highlightColor;
  final Color splashColor;

  // The controller connected to the InputField
  final TextEditingController controller;

  // height and width of each key
  final double height;
  final double width;
  final double spacing;

  // the additional key that can be added to the keyboard
  final bool enableSpacebar;
  final bool enableBackSpace;

  // Additional functionality for the keys
  // Makes the keyboard uppercase
  final bool enableUppercaseAll;

  // Long press to write uppercase letters
  final bool enableLongPressUppercase;
  final Icon backspaceIcon;

  BuiltInKeyboard({
    @required this.controller,
    @required this.layoutType,
    this.letterStyle = const TextStyle(fontSize: 25, color: Colors.black),
    this.borderRadius,
    this.color = Colors.deepOrange,
    this.highlightColor,
    this.splashColor,
    this.height = 46.0,
    this.width = 35.0,
    this.spacing = 8.0,
    this.enableBackSpace = true,
    this.enableUppercaseAll = false,
    this.enableLongPressUppercase = false,
    this.enableSpacebar = false,
    this.backspaceIcon = const Icon(Icons.backspace_outlined, size: 26),
  });
  @override
  BuiltInKeyboardState createState() => BuiltInKeyboardState();
}

class BuiltInKeyboardState extends State<BuiltInKeyboard> {
  @override
  Widget build(BuildContext context) {
    List<Widget> keyboardLayout = layout(widget.layoutType);
    return Column(
      children: [
        Wrap(
          alignment: WrapAlignment.center,
          spacing: 5,
          runSpacing: 5,
          children: keyboardLayout.sublist(0, 10),
        ),
        SizedBox(
          height: widget.spacing,
        ),
        Wrap(
          alignment: WrapAlignment.center,
          spacing: 5,
          runSpacing: 5,
          children: keyboardLayout.sublist(10, 19),
        ),
        SizedBox(
          height: widget.spacing,
        ),
        Wrap(
          alignment: WrapAlignment.center,
          spacing: 5,
          runSpacing: 5,
          children: keyboardLayout.sublist(19),
        ),
        widget.enableSpacebar
            ? Column(
                children: [
                  SizedBox(
                    height: widget.spacing,
                  ),
                  spaceBar(),
                ],
              )
            : SizedBox(),
      ],
    );
  }

  Widget buttonLetter(String letter) {
    return Container(
      height: widget.height,
      width: widget.width,
      child: Material(
        type: MaterialType.button,
        color: widget.color,
        borderRadius: widget.borderRadius,
        child: InkWell(
          highlightColor: widget.highlightColor,
          splashColor: widget.splashColor,
          onTap: () {
            widget.controller.text += letter;
            widget.controller.selection = TextSelection.fromPosition(
                TextPosition(offset: widget.controller.text.length));
          },
          onLongPress: () {
            if (widget.enableLongPressUppercase && !widget.enableUppercaseAll) {
              widget.controller.text += letter.toUpperCase();
              widget.controller.selection = TextSelection.fromPosition(
                  TextPosition(offset: widget.controller.text.length));
            }
          },
          child: Center(
            child: Text(
              letter,
              style: widget.letterStyle,
            ),
          ),
        ),
      ),
    );
  }

  Widget spaceBar() {
    return Container(
      height: widget.height,
      width: widget.width + 160,
      child: Material(
        type: MaterialType.button,
        color: widget.color,
        borderRadius: widget.borderRadius,
        child: InkWell(
          highlightColor: widget.highlightColor,
          splashColor: widget.splashColor,
          onTap: () {
            widget.controller.text += ' ';
            widget.controller.selection = TextSelection.fromPosition(
                TextPosition(offset: widget.controller.text.length));
          },
          child: Center(
            child: Text(
              '_________',
              style: widget.letterStyle,
            ),
          ),
        ),
      ),
    );
  }

  Widget backSpace() {
    return Container(
      height: widget.height,
      width: widget.width + 20,
      child: Material(
        type: MaterialType.button,
        color: widget.color,
        borderRadius: widget.borderRadius,
        child: InkWell(
          highlightColor: widget.highlightColor,
          splashColor: widget.splashColor,
          onTap: () {
            if (widget.controller.text.isNotEmpty) {
              widget.controller.text = widget.controller.text
                  .substring(0, widget.controller.text.length - 1);
              widget.controller.selection = TextSelection.fromPosition(
                  TextPosition(offset: widget.controller.text.length));
            }
          },
          onLongPress: () {
            if (widget.controller.text.isNotEmpty) {
              widget.controller.text = '';
              widget.controller.selection = TextSelection.fromPosition(
                  TextPosition(offset: widget.controller.text.length));
            }
          },
          child: Center(
            child: widget.backspaceIcon,
          ),
        ),
      ),
    );
  }

  List<Widget> layout(String layoutType) {
    List<String> letters = [];
    if (layoutType == 'EN') {
      if (widget.enableUppercaseAll) {
        letters = 'qwertyuiopasdfghjklzxcvbnm'.toUpperCase().split("");
      } else {
        letters = 'qwertyuiopasdfghjklzxcvbnm'.split("");
      }
    } else if (layoutType == 'FR') {
      if (widget.enableUppercaseAll) {
        letters = 'azertyuiopqsdfghjklmwxcvbn'.toUpperCase().split("");
      } else {
        letters = 'azertyuiopqsdfghjklmwxcvbn'.split("");
      }
    }

    List<Widget> keyboard = [];
    letters.forEach((String letter) {
      keyboard.add(
        buttonLetter(letter),
      );
    });
    if (widget.enableBackSpace) {
      keyboard.add(backSpace());
    }
    return keyboard;
  }
}


lib/built_in_keyboard.dart
library built_in_keyboard;

import 'dart:io';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import 'language.dart';

class BuiltInKeyboard extends StatefulWidget {
  // Language of the keyboard
  final Language language;

  // layout of the keyboard
  final Layout layout;

  // The controller connected to the InputField
  final TextEditingController? controller;

  // Vertical spacing between key rows
  final double spacing;

  // Border radius of the keys
  final BorderRadius? borderRadius;

  // Color of the keys
  final Color color;

  // TextStyle of the letters in the keys (fontsize, fontface)
  final TextStyle letterStyle;

  // the additional key that can be added to the keyboard
  final bool enableSpaceBar;
  final bool enableBackSpace;
  final bool enableCapsLock;

  // height and width of each key
  final double? height;
  final double? width;

  // Additional functionality for the keys //

  // Makes the keyboard uppercase
  final bool enableAllUppercase;

  // Long press to write uppercase letters
  final bool enableLongPressUppercase;

  // The color displayed when the key is pressed
  final Color? highlightColor;
  final Color? splashColor;

  BuiltInKeyboard({
    @required this.controller,
    this.language = Language.EN,
    this.layout = Layout.QWERTY,
    this.height,
    this.width,
    this.spacing = 8.0,
    this.borderRadius,
    this.color = Colors.deepOrange,
    this.letterStyle = const TextStyle(fontSize: 25, color: Colors.black),
    this.enableSpaceBar = false,
    this.enableBackSpace = true,
    this.enableCapsLock = false,
    this.enableAllUppercase = false,
    this.enableLongPressUppercase = false,
    this.highlightColor,
    this.splashColor,
  });
  @override
  BuiltInKeyboardState createState() => BuiltInKeyboardState();
}

class BuiltInKeyboardState extends State<BuiltInKeyboard> {
  double? height;
  double? width;
  bool capsLockUppercase = false;
  @override
  Widget build(BuildContext context) {
    double screenWidth = MediaQuery.of(context).size.width;
    double screenHeight = MediaQuery.of(context).size.height;
    height = screenHeight > 800 ? screenHeight * 0.059 : screenHeight * 0.07;
    width = screenWidth > 350 ? screenWidth * 0.084 : screenWidth * 0.082;
    List<Widget> keyboardLayout = layout();
    double hspacing;
    int topLen, midLen;
    try {
      hspacing = double.parse(languageConfig[widget.language]![widget.layout]![
          'horizontalSpacing']!);
      topLen = int.parse(
          languageConfig[widget.language]![widget.layout]!['topLength']!);
      midLen = int.parse(
          languageConfig[widget.language]![widget.layout]!['middleLength']!);
    } catch (_CastError) {
      printError(
          "Uknown language or layout was used, or Incorrect combination of language-layout");
      exit(0);
    }
    return Column(
      children: [
        Wrap(
          alignment: WrapAlignment.center,
          spacing: hspacing,
          runSpacing: 5,
          children: keyboardLayout.sublist(0, topLen),
        ),
        SizedBox(
          height: widget.spacing,
        ),
        Wrap(
          alignment: WrapAlignment.center,
          spacing: hspacing,
          runSpacing: 5,
          children: keyboardLayout.sublist(topLen, topLen + midLen),
        ),
        SizedBox(
          height: widget.spacing,
        ),
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: [
            widget.enableCapsLock
                ? capsLock()
                : SizedBox(
                    width: (widget.width ?? width)! + 20,
                  ),
            Wrap(
              alignment: WrapAlignment.center,
              spacing: hspacing,
              runSpacing: 5,
              children: keyboardLayout.sublist(topLen + midLen),
            ),
            widget.enableBackSpace
                ? backSpace()
                : SizedBox(
                    width: (widget.width ?? width)! + 20,
                  ),
          ],
        ),
        widget.enableSpaceBar
            ? Column(
                children: [
                  SizedBox(
                    height: widget.spacing,
                  ),
                  spaceBar(),
                ],
              )
            : SizedBox(),
      ],
    );
  }

  // Letter button widget
  Widget buttonLetter(String letter) {
    return ClipRRect(
      borderRadius: widget.borderRadius ?? BorderRadius.circular(0),
      child: Container(
        height: widget.height ?? height,
        width: widget.width ?? width,
        child: Material(
          type: MaterialType.button,
          color: widget.color,
          child: InkWell(
            highlightColor: widget.highlightColor,
            splashColor: widget.splashColor,
            onTap: () {
              HapticFeedback.heavyImpact();
              widget.controller?.text += letter;
              widget.controller?.selection = TextSelection.fromPosition(
                  TextPosition(offset: widget.controller!.text.length));
            },
            onLongPress: () {
              if (widget.enableLongPressUppercase &&
                  !widget.enableAllUppercase) {
                widget.controller?.text += letter.toUpperCase();
                widget.controller?.selection = TextSelection.fromPosition(
                    TextPosition(offset: widget.controller!.text.length));
              }
            },
            child: Center(
              child: Text(
                letter,
                style: widget.letterStyle,
              ),
            ),
          ),
        ),
      ),
    );
  }

  // Spacebar button widget
  Widget spaceBar() {
    return ClipRRect(
      borderRadius: widget.borderRadius ?? BorderRadius.circular(0),
      child: Container(
        height: widget.height ?? height,
        width: (widget.width ?? width)! + 160,
        child: Material(
          type: MaterialType.button,
          color: widget.color,
          child: InkWell(
            highlightColor: widget.highlightColor,
            splashColor: widget.splashColor,
            onTap: () {
              HapticFeedback.heavyImpact();
              widget.controller?.text += ' ';
              widget.controller?.selection = TextSelection.fromPosition(
                  TextPosition(offset: widget.controller!.text.length));
            },
            child: Center(
              child: Text(
                '_________',
                style: TextStyle(
                  fontSize: widget.letterStyle.fontSize,
                  color: widget.letterStyle.color,
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }

  // Backspace button widget
  Widget backSpace() {
    return ClipRRect(
      borderRadius: widget.borderRadius ?? BorderRadius.circular(0),
      child: Container(
        height: widget.height ?? height,
        width: (widget.width ?? width)! + 20,
        child: Material(
          type: MaterialType.button,
          color: widget.color,
          child: InkWell(
            highlightColor: widget.highlightColor,
            splashColor: widget.splashColor,
            onTap: () {
              HapticFeedback.heavyImpact();
              if (widget.controller!.text.isNotEmpty) {
                widget.controller?.text = widget.controller!.text
                    .substring(0, widget.controller!.text.length - 1);
                widget.controller?.selection = TextSelection.fromPosition(
                    TextPosition(offset: widget.controller!.text.length));
              }
            },
            onLongPress: () {
              if (widget.controller!.text.isNotEmpty) {
                widget.controller?.text = '';
                widget.controller?.selection = TextSelection.fromPosition(
                    TextPosition(offset: widget.controller!.text.length));
              }
            },
            child: Center(
              child: Icon(
                Icons.backspace,
                size: widget.letterStyle.fontSize,
                color: widget.letterStyle.color,
              ),
            ),
          ),
        ),
      ),
    );
  }

  // Capslock button widget
  Widget capsLock() {
    return ClipRRect(
      borderRadius: widget.borderRadius ?? BorderRadius.circular(0),
      child: Container(
        height: widget.height ?? height,
        width: (widget.width ?? width)! + 20,
        child: Material(
          type: MaterialType.button,
          color: widget.color,
          child: InkWell(
            highlightColor: widget.highlightColor,
            splashColor: widget.splashColor,
            onTap: () {
              HapticFeedback.heavyImpact();
              setState(() {
                capsLockUppercase = !capsLockUppercase;
              });
            },
            child: Center(
              child: Icon(
                Icons.keyboard_capslock,
                size: widget.letterStyle.fontSize,
                color: widget.letterStyle.color,
              ),
            ),
          ),
        ),
      ),
    );
  }

  // Keyboard layout list
  List<Widget> layout() {
    List<String> letters = [];
    try {
      letters =
          languageConfig[widget.language]![widget.layout]!['layout']!.split("");
    } catch (_CastError) {
      printError(
          "Uknown language or layout was used, or Incorrect combination of language-layout");
      exit(0);
    }

    List<Widget> keyboard = [];
    letters.forEach((String letter) {
      keyboard.add(
        buttonLetter(letter),
      );
    });
    return keyboard;
  }
}

void printError(String text) {
  print('\x1B[31m$text\x1B[0m');
}
